### ISSUE TITLE
non-convergence since upgrade

### ISSUE THREAD

**Issue by stanbiryukov** [cornellius-gp/gpytorch#155]
Hi all, 
I upgraded gpytorch to test out the new white noise kernel and now a 3D-exact GP classification model (without white noise addition) that had previously converged is failing.  My target data is -1/1.  Using the RBF kernel, I see the following error:

`File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/module.py", line 162, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/mlls/variational_marginal_log_likelihood.py", line 29, in forward
    variational_strategy.kl_divergence().sum() for variational_strategy in self.model.variational_strategies()
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/mlls/variational_marginal_log_likelihood.py", line 29, in <genexpr>
    variational_strategy.kl_divergence().sum() for variational_strategy in self.model.variational_strategies()
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/variational/mvn_variational_strategy.py", line 27, in kl_divergence
    inv_quad_rhs=inv_quad_rhs, log_det=True
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/lazy/lazy_variable.py", line 454, in inv_quad_log_det
    preconditioner=self._preconditioner()[0],
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/lazy/added_diag_lazy_variable.py", line 50, in _preconditioner
    self._piv_chol_self = pivoted_cholesky.pivoted_cholesky(self._lazy_var, max_iter)
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/utils/pivoted_cholesky.py", line 97, in pivoted_cholesky
    return L[0, :m, :]
RuntimeError: dimension out of range (expected to be in range of [-1, 0], but got 1)`

And with Matern 2.5:
`  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/models/variational_gp.py", line 37, in __call__
    prior_output = self.prior_output()
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/models/abstract_variational_gp.py", line 65, in prior_output
    res = GaussianRandomVariable(res.mean(), res.covar().evaluate_kernel())
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/lazy/lazy_variable.py", line 274, in evaluate_kernel
    return self.representation_tree()(*self.representation())
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/lazy/lazy_variable.py", line 578, in representation_tree
    return LazyVariableRepresentationTree(self)
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/lazy/lazy_variable_representation_tree.py", line 14, in __init__
    for arg in lazy_var._args:
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/lazy/mul_lazy_variable.py", line 73, in _args
    right_lazy_var = RootLazyVariable(right_lazy_var.root_decomposition())
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/lazy/lazy_variable.py", line 593, in root_decomposition
    )(*self.representation())
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/functions/_root_decomposition.py", line 54, in forward
    eigenvalues, eigenvectors = lanczos_tridiag_to_diag(t_mat)
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/utils/lanczos.py", line 170, in lanczos_tridiag_to_diag
    return batch_symeig(t_mat)
  File "/Users/stanleybiryukov/miniconda2/envs/idp/lib/python2.7/site-packages/gpytorch/utils/eig.py", line 23, in batch_symeig
    evals, evecs = mat[i, j].symeig(eigenvectors=True)
RuntimeError: Lapack Error syev : 2 off-diagonal elements didn't converge to zero at /Users/soumith/minicondabuild3/conda-bld/pytorch_1524587833086/work/aten/src/TH/generic/THTensorLapack.c:388`

Please let me know if I can provide additional details to help with de-bugging.

**Comment by jacobrgardner** [comment]
I'll take a look. Do you have simple example data that produces either problem? 

Both errors look like problems I'd expect to encounter when faced with bad kernel matrices due to bad hyperparameters (e.g., the kernel matrix is all ones because the lengthscale is very large relative to the Euclidean distances encountered in the data). Can you print out the kernel matrix (something like `print(covar_x.evaluate())`?

**Comment by stanbiryukov** [comment]
Thanks, Jacob.  Here's example dummy data that reproduces the error:
[train_numpy_arrays.zip](https://github.com/cornellius-gp/gpytorch/files/2066194/train_numpy_arrays.zip)

And the kernel matrix:
`Iter 21/200 - Loss: 119.710
tensor([[ 1.2208e+00,  9.6203e-01,  6.2319e-01,  ...,  1.9692e-02,
          8.6667e-03,  3.5650e-03],
        [ 9.6203e-01,  1.2208e+00,  1.0427e+00,  ...,  4.0918e-02,
          1.9692e-02,  8.6667e-03],
        [ 6.2319e-01,  1.0427e+00,  1.2208e+00,  ...,  6.7531e-02,
          3.5650e-02,  1.6817e-02],
        ...,
        [ 1.9692e-02,  4.0918e-02,  6.7531e-02,  ...,  1.2208e+00,
          9.6203e-01,  5.4698e-01],
        [ 8.6667e-03,  1.9692e-02,  3.5650e-02,  ...,  9.6203e-01,
          1.2208e+00,  9.6203e-01],
        [ 3.5650e-03,  8.6667e-03,  1.6817e-02,  ...,  5.4698e-01,
          9.6203e-01,  1.2208e+00]])
Iter 22/200 - Loss: 116.482
tensor([[ 1.2312e+00,  9.6663e-01,  6.2242e-01,  ...,  1.8906e-02,
          8.2499e-03,  3.3629e-03],
        [ 9.6663e-01,  1.2312e+00,  1.0490e+00,  ...,  3.9591e-02,
          1.8906e-02,  8.2499e-03],
        [ 6.2242e-01,  1.0490e+00,  1.2312e+00,  ...,  6.5699e-02,
          3.4443e-02,  1.6119e-02],
        ...,
        [ 1.8906e-02,  3.9591e-02,  6.5699e-02,  ...,  1.2312e+00,
          9.6663e-01,  5.4537e-01],
        [ 8.2499e-03,  1.8906e-02,  3.4443e-02,  ...,  9.6663e-01,
          1.2312e+00,  9.6663e-01],
        [ 3.3629e-03,  8.2499e-03,  1.6119e-02,  ...,  5.4537e-01,
          9.6663e-01,  1.2312e+00]])
Iter 23/200 - Loss: 105.824
tensor([[nan., nan., nan.,  ..., nan., nan., nan.],
        [nan., nan., nan.,  ..., nan., nan., nan.],
        [nan., nan., nan.,  ..., nan., nan., nan.],
        ...,
        [nan., nan., nan.,  ..., nan., nan., nan.],
        [nan., nan., nan.,  ..., nan., nan., nan.],
        [nan., nan., nan.,  ..., nan., nan., nan.]])`

**Comment by jacobrgardner** [comment]
Okay, the problem was really silly. In the simple GP classification notebook, we were doing:
```python
covar_x = covar_x.mul(self.log_outputscale.exp().expand_as(covar_x))
```
But because of some recent changes internally, this will result in trying to take the square root of an all constant matrix (`self.log_outputscale.exp().expand_as(covar_x)`), which results in NaNs. This needs to be:
```python
covar_x = covar_x.mul(self.log_outputscale.exp())
```
which is how outputscales are handled everywhere else (including the unit tests, which is how this slipped by). I've updated the simple GP classification notebook in the repo. With this change, I was able to run on your supplied data with no issue.

Hope this helps!

(@stanbiryukov)
